common/src/deposits.rs

```diff
@@ -123,84 +281,65 @@ impl DepositScriptInputs {
-
-/// This function checks that the deposit script is valid. Specifically, it
-/// checks that it follows the format laid out in
-/// https://github.com/stacks-network/sbtc/issues/30, where the script is
-/// expected to be
-/// ```text
-///  <deposit-data> OP_DROP OP_PUSHBYTES_32 <x-only-public-key> OP_CHECKSIG
-/// ```
-pub fn parse_deposit_script(deposit_script: &ScriptBuf) -> Result<DepositScriptInputs, Error> {
-    let script = deposit_script.as_bytes();
 
-    // Valid deposit scripts cannot be less than this length.
-    if script.len() < STANDARD_SCRIPT_LENGTH {
-        return Err(Error::InvalidDepositScript);
+    /// This function checks that the deposit script is valid.
+    /// Specifically, it checks that it follows the format laid out in
+    /// https://github.com/stacks-network/sbtc/issues/30, where the script
+    /// is expected to be
+    /// ```text
+    ///  <deposit-data> OP_DROP OP_PUSHBYTES_32 <x-only-public-key> OP_CHECKSIG
+    /// ```
+    pub fn parse(deposit_script: &ScriptBuf) -> Result<Self, Error> {
+        let script = deposit_script.as_bytes();
+
+        // Valid deposit scripts cannot be less than this length.
+        if script.len() < STANDARD_SCRIPT_LENGTH {
+            return Err(Error::InvalidDepositScript);
+        }
+        // This cannot panic because of the above check and the fact that
+        // DEPOSIT_SCRIPT_FIXED_LENGTH < STANDARD_SCRIPT_LENGTH.
+        let (params, check) = script.split_at(script.len() - DEPOSIT_SCRIPT_FIXED_LENGTH);
+        // Below, we know the script length is DEPOSIT_SCRIPT_FIXED_LENGTH,
+        // because of how `slice::split_at` works, so we know the
+        // public_key variable has length 32.
+        let [OP_DROP, 32, public_key @ .., OP_CHECKSIG] = check else {
+            return Err(Error::InvalidDepositScript);
+        };
+        // In bitcoin script, the code for pushing N bytes onto the stack
+        // is OP_PUSHBYTES_N where N is between 1 and 75 inclusive. The
+        // byte representation of these opcodes is the byte representation
+        // of N. If you need to push between 76 and 255 bytes of data then
+        // you need to use the OP_PUSHDATA1 opcode (you can also use this
+        // opcode to push between 1 and 75 bytes on the stack, but it's
+        // cheaper to use the OP_PUSHBYTES_N opcodes when you can). When
+        // need to check all cases since contract addresses can have a size
+        // of up to 151 bytes.
+        let data = match params {
+            // This branch represents a contract address.
+            [OP_PUSHDATA1, n, data @ ..] if data.len() == *n as usize && *n < 160 => data,
+            // This branch can be a standard (non-contract) Stacks
+            // addresses when n == 30 and is a contract address otherwise.
+            [n, data @ ..] if data.len() == *n as usize && *n < 76 => data,
+            _ => return Err(Error::InvalidDepositScript),
+        };
```

Comment:
Can you describe what the rust syntax `data @ ..` is doing? I've also never seen this match syntax, it's interesting. Also, why `n < 160`, if contract addresses can be up to 151 bytes?

Reply:
Ah, per [this stacks-core code](https://github.com/stacks-network/stacks-core/blob/202b45cc65b88a15d31a33370e47850cf98d7c21/clarity/src/vm/representations.rs#L81), it seems that the max size of a contract name is 128 bytes, so 128 + 30 = 158, and 1 byte for the prefix, which results in a max of 159. I get the "why n < 160" now.

Reply:
But now I'm realizing that `data` also includes the max fee (8 bytes), so does that make the max size 167?

Reply:
> But now I'm realizing that `data` also includes the max fee (8 bytes), so does that make the max size 167?

Hmmm, I don't think so. For the principal, I thought it was 1 byte for the type prefix, 21 bytes for the standard address, 1 byte for the contract name, and 128 bytes for the contract name, for a total of 151 bytes. Then there is 8 bytes for the max fee, bringing it to 159 bytes. Oh, I need to update the comment to note that it includes the max fee!

Anyway, does that make sense? And thanks for looking this part over again.

Reply:
I'm going to add a test for the max length contract parse.

Reply:
Yeah that makes sense, when I read your comment here "when n == 30 it's standard" I just assumed the address part was 30 bytes and didn't think about it. Yeah, it's 1 byte version, 20 byte pubkey hash.

Math checks out! For future proofing, can you add a test case where a contract address is maxed out? Maybe just change the `CONTRACT_ADDRESS` var to be max length?

Reply:
Do you think you could add a comment somewhere in this file outlining the "wire format" for `data`? I find [this kind of format](https://github.com/stacks-network/stacks-core/blob/202b45cc65b88a15d31a33370e47850cf98d7c21/stackslib/src/chainstate/burn/operations/stack_stx.rs#L188-L191) helpful:

```
0      2  3                             19           20                  53                 69                        73
|------|--|-----------------------------|------------|-------------------|-------------------|-------------------------|
magic  op         uSTX to lock (u128)     cycles (u8)     signer key (optional)   max_amount (optional u128)  auth_id (optional u32)
```

Reply:
> Do you think you could add a comment somewhere in this file outlining the "wire format" for `data`?

Oh yeah, I find it very useful. Good call, I'll add that here.

> Math checks out! For future proofing, can you add a test case where a contract address is maxed out? Maybe just change the `CONTRACT_ADDRESS` var to be max length?

Not a problem, added one here https://github.com/stacks-network/sbtc/pull/350/commits/63b46c04ca17e1f877a55851e0b9d97b3b9bf970. I think this is what you had in mind, I wanted to add one and just forgot.

Reply:
Added a wire format diagram thingy in https://github.com/stacks-network/sbtc/pull/350/commits/a0f146e9647d6776fc6751ac669f4c75d84eeac2.